### ISSUE TITLE
getBayerBinaryMessage: waiting for message 0x81, got 0x80 Timeout waiting for message

### ISSUE THREAD

**Issue by benfiscal** [pazaan/decoding-contour-next-link#38]
Hi,
I'm running the command `sudo python -m decoding-contour-next-link.read_minimed_next24
`from my pi0 rig. I am successful getting the pump history, but when it's going to fetch the sensor history info I receive the following:

```
Getting sensor history info
2018-11-06 21:51:51,748 WARNING [__main__] ## getBayerBinaryMessage: waiting for message 0x81, got 0x80
2018-11-06 21:52:01,098 WARNING [__main__] ## getBayerBinaryMessage: waiting for message 0x81, got 0x80
2018-11-06 21:52:11,102 WARNING [__main__] Timeout waiting for message
2018-11-06 21:52:11,106 ERROR [__main__] Unexpected error in client downloadOperations
Traceback (most recent call last):
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 1249, in downloadPumpSession
    downloadOperations(mt)
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 1300, in pumpDownload
    sensHistoryInfo = mt.getPumpHistoryInfo(start_date, datetime.datetime.max, HISTORY_DATA_TYPE.SENSOR_DATA)
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 1002, in getPumpHistoryInfo
    self.getBayerBinaryMessage(0x81) # Read the 0x81
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 956, in getBayerBinaryMessage
    message = BayerBinaryMessage.decode(self.readMessage())
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 783, in readMessage
    raise TimeoutException( 'Timeout waiting for message' )
TimeoutException: Timeout waiting for message
2018-11-06 21:52:22,133 WARNING [__main__] Timeout waiting for message
2018-11-06 21:52:34,142 WARNING [__main__] Timeout waiting for message
2018-11-06 21:52:34,146 WARNING [__main__] Unexpected error by requestCloseConnection, ignoring
Traceback (most recent call last):
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 878, in closeConnection
    self.readMessage()
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 783, in readMessage
    raise TimeoutException( 'Timeout waiting for message' )
TimeoutException: Timeout waiting for message
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 1333, in <module>
    downloadPumpSession(pumpDownload)
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 1249, in downloadPumpSession
    downloadOperations(mt)
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 1300, in pumpDownload
    sensHistoryInfo = mt.getPumpHistoryInfo(start_date, datetime.datetime.max, HISTORY_DATA_TYPE.SENSOR_DATA)
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 1002, in getPumpHistoryInfo
    self.getBayerBinaryMessage(0x81) # Read the 0x81
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 956, in getBayerBinaryMessage
    message = BayerBinaryMessage.decode(self.readMessage())
  File "/home/pi/decoding-contour-next-link/read_minimed_next24.py", line 783, in readMessage
    raise TimeoutException( 'Timeout waiting for message' )
__main__.TimeoutException: Timeout waiting for message
```

**Comment by mortlind** [comment]
I get these `TimeoutException`s occationally, but quite frequently. Do you get them consistently or occasionally. I suspect it to be a power issue with the Contour Next Link device, since I seem to get them often on an RPi, but less frequently on my laptop.

**Comment by mortlind** [comment]
Another thing I believe to have noticed is that the power issue may be provoked by issuing downloads in immediate succession. Hence I leave a time lapse between getting the status, getting the pump event history, and getting the sensor event history. Are you acquiring the sensor event history immediately after having acquired the pump event history?

**Comment by benfiscal** [comment]
Yes, I just run `sudo python -m decoding-contour-next-link.read_minimed_next24` without any options or flags so it acquires the sensor event history immediately after having acquired the pump event history.  I guess that there is a way to specify options. Do you know if this is documented somewhere?

**Comment by mortlind** [comment]
I think the default application in the module is just an example, and "selfdocumenting". You could try introduce a `sleep` of a couple of seconds between the acquisitions. But you could also just test the same application from a laptop. I suspect the USB power on the RPi is weak, especially if it is an older model, and definitely if you use a 1 A supply for the RPi itself.

**Comment by benfiscal** [comment]
I'll try your suggestions. I looked at the code in `read_minimed_next24.py` and see where I could insert that `sleep`. Right now I'm mainly interested in getting the sensor information, so I could try to omitt the pup history since I would like to use the 640G as my cgm source in setting up openAPS with a loopable pump.

**Comment by pazaan** [comment]
Admittedly, the python code is well out of date with what's in the Nightscout Uploader, or in Tidepool (including all the microsleeps)

**Comment by mortlind** [comment]
@pazaan Could you elaborate on that? I am not familiar with Tidepool or the Nightscout Uploader. Are they also Python-based controllers for the CNL?

**Comment by pazaan** [comment]
This repo was my original proof-of-concept when I was reverse engineering the 600-series device comms.
Since then, the [600-series Nightscout Uploader](https://github.com/pazaan/600SeriesAndroidUploader/) (in Java) has been created, as has the [Tidepool 600-series device driver](https://github.com/tidepool-org/chrome-uploader/tree/master/lib/drivers/medtronic600) (in JavaScript).
After having written those implementations (as well as the additional comms enhancements), I haven't back-ported those findings into this Python codebase.

**Comment by mortlind** [comment]
@pazaan Thanks for that information. It is really nice with some background knowledge. I would be interested in looking into backporting the enhancements from the Nightscout and the Tidepool bases. But since the current code base here actually works well enough for me, using it every day, I may struggle to find enough motivation :-)

**Comment by benfiscal** [comment]
I would also be interested in looking into backporting those enhancements, but since my knowledge of both Java and Python is limited, the task might (at least in the short term) be insurmountable.

**Comment by christopher-john-czettel** [comment]
I would also like to offer my help in backporting the new findings about the protocol into this codebase. Would you maybe care to give me/us some pointers on what (or where) you basically enhanced in your java/javascript protocol implementation? :-)

**Comment by pazaan** [comment]
> Would you maybe care to give me/us some pointers on what (or where) you basically enhanced in your java/javascript protocol implementation? :-)

@christopher-john-czettel The comms is carried out in [the medtronic600Driver.js](https://github.com/tidepool-org/chrome-uploader/blob/master/lib/drivers/medtronic600/medtronic600Driver.js) in the Tidepool Uploader. Since there have been a lot of changes, I don't think that it's as easy as pointing out a simple delta.

**Comment by tuzoenduro** [comment]
Hi Lennart, Christopher,

if you have this running on a Rasp 0, does this mean that openAps could be run on the same device and thus skip adding the uploader phone to the mix?

At the moment I'm trying to close loop using the 600 series uploader to send the readings (from my 640) to nightscout and then have the PI 0 pick up from NS and apply the algorithms to a 722 pump.

If I could bypass the need for an uploader phone and just plug the CNL to the PI this would lighten the load of electronics I carry with me. Thank you for your help and insight.

**Comment by pazaan** [comment]
@tuzoenduro - yes, one could run this on the same RaspPi as on OpenAPS, and skip adding the uploader phone.

**Comment by tuzoenduro** [comment]
@pazaan one question, for the moment the /decoding contour next link only recuperates the BG values? this is pretty much all I would need as the rest is done directly by the openAPS side of things.

Do the BG values get sent to NS? or is there a way to share them with the openAPS soft? My coding knowledge is very good for following instructions but does not go beyond that so I would need a little help here.

Thanks for all your support.

**Comment by pazaan** [comment]
This python codebase doesn't do anything other than pull data from the pump. At this stage, it doesn't send the data to Nightscout (see https://github.com/pazaan/decoding-contour-next-link/issues/7), or any other shared data repository.

**Comment by ondrej1024** [comment]
I have started backporting the improvements of the more up to date Android uploader code to the Python driver. See [here](https://github.com/pazaan/decoding-contour-next-link/issues/46#issuecomment-607238572) for details.
